src/firmware/app/control/trajectory_planner.h

```diff
@@ -11,293 +13,434 @@ typedef struct FirmwareRobotPathParameters
 
 typedef struct PositionTrajectory
 {
-    PositionTrajectoryElement_t* trajectory_elements;
-    FirmwareRobotPathParameters_t path_parameters;
-    float* speed_profile;
+    float x_position[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float y_position[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float orientation[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float linear_speed[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float angular_speed[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float time_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
 } PositionTrajectory_t;
 
-// Struct that defines a single point on a velocity trajectory
-// Includes the velocity and Time data corresponding to that point
-typedef struct VelocityTrajectoryElement
-{
-    Vector2d_t linear_velocity;
-    float angular_velocity;
-    float time;
-} VelocityTrajectoryElement_t;
-
 typedef struct VelocityTrajectory
 {
-    VelocityTrajectoryElement_t* trajectory_elements;
-    FirmwareRobotPathParameters_t path_parameters;
+    float x_velocity[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float y_velocity[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float angular_velocity[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float time_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
 } VelocityTrajectory_t;
 
 typedef enum TrajectoryPlannerGenerationStatus
 {
     OK,
     FINAL_VELOCITY_TOO_HIGH,
     INITIAL_VELOCITY_TOO_HIGH,
-
     // INTERPOLATION_ELEMENT_MAXED_OUT is returned when the constant time interpolation
-    // function uses up all the avaiable array space provided to it. This can happen
+    // function uses up all the available array space provided to it. This can happen
     // because constant interpolation period trajectories do not necessarily have the same
-    // number of elements as their constant arc-length counterparts
+    // number of elements as their constant parameterization counterparts
     INTERPOLATION_ELEMENT_MAXED_OUT,
 } TrajectoryPlannerGenerationStatus_t;
 
 /**
- * Returns a planned trajectory with the list of guarantees based on the assumptions below
+ * Function generates the segment lengths and the states at each node along the specified
+ * constant-parameterization segments
+ *
+ * @pre All arrays must be pre-allocated up to TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS
+ *
+ * @param t_start [in] The starting parameterization value
+ *
+ * @param t_end  [in] The ending parameterization value
+ *
+ * @param poly  [in] The 1d polynomial representing the path along the trajectory
+ *
+ * @param segment_lengths [out] The length of segments between each node specified by the
+ * constant parameterization
+ *
+ * @param node_values [out] The evaluation of the polynomial at each value of the constant
+ * parameterization
+ *
+ * @param num_elements [in] The number of elements (nodes) to be in the trajectory
+ *
+ * Node1  |--segment length---| Node2
+ *   ->   *-------------------*  <-
+ */
+void app_trajectory_planner_generateSegmentNodesAndLengths(
```

Comment:
How does this compare to the linear version of the function of the same name? Is it angular? angular and linear?

Reply:
Ah yes, I could probably clarify this.

This one is generic and takes a 1d polynomial - which doesn't work for the 2d linear profile. The 'linear' version is a wrapper of running the 1d version twice then putting together the x and y segment lengths.

Reply:
maybe add 1d to the function name

Reply:
Changed function to 1d and 2d